src/components/AddressDetailExplorer.js

```diff
@@ -46,412 +49,362 @@ class AddressDetailExplorer extends React.Component {
-              // User already have a token selected on the URL
-              this.setState({ selectedToken: queryParams.token }, () => {
-                this.reloadData();
-              });
-            } else {
-              this.reloadData();
+      /**
+       * Check if the searched address is on the inputs or outputs of the new tx
+       *
+       * @param {Object} tx Transaction data received in the websocket
+       * @param {boolean} checkToken If should also check if token is the same, or just address
+       *
+       * @return {boolean} True if should update the list, false otherwise
+       */
+      function shouldUpdate(tx, checkToken) {
+        const arr = [...tx.outputs, ...tx.inputs];
+        const { token: queryToken } = pagination.current.obtainQueryParams();
+
+        for (const element of arr) {
+          if (element.decoded.address === address) {
+            // Address is the same
+            if ((checkToken && element.token === queryToken) || !checkToken) {
+              // Need to check token and token is the same, or no need to check token
+              return true;
             }
           }
-        );
-      } else {
-        this.setState({ errorMessage: 'Invalid address.' });
+        }
+
+        return false;
       }
-    });
-  };
+    },
+    [address]
+  );
+
+  // Initialization effect for every address change
+  useEffect(() => {
+    // Clear all optional params since the address on the URL has changed
+    pagination.current.clearOptionalQueryParams();
+
+    // Validate the new address before executing any queries
+    const network = hathorLib.config.getNetwork();
+    const addressObj = new hathorLib.Address(address, { network });
+    if (!addressObj.isValid()) {
+      setErrorMessage('Invalid address.');
+      // No listener was registered, so no unmount function is necessary
+      return undefined;
+    }
+
+    // Trigger effect that fetches all tokens for this address and loads one of them on screen
+    setLoadingTokens(true);
+
+    // Starting the Websocket
+    WebSocketHandler.on('network', handleWebsocket);
+    // Return a function to remove the Websocket listener
+    return () => {
+      WebSocketHandler.removeListener('network', handleWebsocket);
+    };
+  }, [handleWebsocket, address]);
 
   /**
    * Update transactions data state after requesting data from the server
    *
-   * @param {Object} queryParams URL parameters
+   * @param addressToFetch
+   * @param tokenToFetch
+   * @param [lastTx]
+   * @param [lastTs]
    */
-  getHistoryData = (lastTx, lastTs) => {
-    return addressApi
-      .getHistory(this.state.address, this.state.selectedToken, TX_COUNT, lastTx, lastTs)
-      .then(response => {
-        if (!response) {
-          // An error happened with the API call
-          this.setState({
-            showReloadTokenButton: true,
-          });
-          return;
-        }
+  const getHistoryData = useCallback(
```

Comment:
This callback requires `txCache` but it is the only one to call `setTxCache`.
Since this updates the `txCache` it I think we should not use setTxCache and just update the txCache directly.

The idea would be that the txCache is never "set", meaning that the same object will be used between re-renders.
This method adds transactions to the object and the next re-render will have the same object.
What do you think?

Reply:
State immutability is [an important concept for React](https://reacttraining.com/blog/state-in-react-is-immutable). By always using `set` and never mutating the object we can benefit from some features like "time machine" when debugging our application.

However, when studying some other alternative impelmentations for this `getHistoryData` based on your comment, I found a bug that could erase some transactions when under race conditions.

Fixed it on https://github.com/HathorNetwork/hathor-explorer/pull/317/commits/f979f479cbac5b4eeabb1c0410af56e6b9469626

Reply:
My comment had nothing to do with react's state immutability,

The idea was that a change in `txCache` does not seem to change any rendered item.
So a change on the `txCache` would not require a re-render of the screen and the same object could be used between re-renders (for instance, we could use a `useMemo` without dependencies).

**But** after reading more I saw that the only place it is used is to check that a transaction only has authorities (on the history list) and this is updated async so it will require the table to be re-rendered.